File: 1. 단계별로_풀어보기/20~29/22. 누적 합/6. 25682.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const [N, M, K] = inputArr[0].trim().split(' ').map(Number)
const board = inputArr.slice(1).map(row => row.trim().split(''))

// W로 시작하는것 기준, 바뀌어야 하는지 여부 배열
const okArr = Array.from({length: N}, () => Array.from({length: M}).fill(0))

// okArr 연산
for (let row = 0; row < N; row++) {
  for (let col = 0; col < M; col++) {
    if ((row + col) % 2 === 0) {
      okArr[row][col] = board[row][col] === 'B' ? 1 : 0
    } // ::
    else {
      okArr[row][col] = board[row][col] === 'W' ? 1 : 0
    }
  }
}

// 누적합 배열 생성
const sumArr = Array.from({length: N + 1}, () => Array.from({length: M + 1}).fill(0))

// 누적합 배열 연산
for (let row = 1; row <= N; row++) {
  for (let col = 1; col <= M; col++) {
    sumArr[row][col] = okArr[row - 1][col - 1] + sumArr[row - 1][col] + sumArr[row][col - 1] - sumArr[row - 1][col - 1]
  }
}

// 결과값 구하기
let result = K * K
for (let row = K; row <= N; row++) {
  for (let col = K; col <= M; col++) {
    const nowSum = sumArr[row][col] - sumArr[row - K][col] - sumArr[row][col - K] + sumArr[row - K][col - K]
    result = Math.min(result, nowSum, K * K - nowSum)
  }
}

// 출력
process.stdout.write(`${result}`)

File: 1. 단계별로_풀어보기/20~29/22. 누적 합/6. 25682.py
# PyPy3 으로 돌려야 시간초과가 안난다

import sys

input = sys.stdin.readline
output = sys.stdout.write

N, M, K = map(int, input().split())
board = list(list(input().strip()) for i in range(N))

# Pythonic 하게 만들어봤다. 근데 가독성이 좀...
okArr = list(list(1 if (i + j) % 2 == 0 and board[i][j] == 'B' or (i + j) % 2 == 1 and board[i][j] == 'W' else 0 for j in range(M)) for i in range(N))


sumArr = list(list(0 for i in range(M + 1)) for j in range(N + 1))

for row in range(1, N + 1):
  for col in range(1, M + 1):
    sumArr[row][col] = okArr[row - 1][col - 1] + sumArr[row - 1][col] + sumArr[row][col - 1] - sumArr[row - 1][col - 1]


# Pythonic 하게 구하려면 코드가 너무 길어진다.
result = K * K
for row in range(K, N + 1):
  for col in range(K, M + 1):
    nowSum = sumArr[row][col] - sumArr[row - K][col] - sumArr[row][col - K] + sumArr[row - K][col - K]
    result = min(result, nowSum, K * K - nowSum)

output(f"{result}")



